### ISSUE TITLE
4.keras_CNN_mnist.ipynb

### ISSUE THREAD

**Issue by Hank666818** [Hank666818/-#3]
from tensorflow import keras
from keras.models import Sequential
from keras.layers import Dense, Conv2D, MaxPooling2D, Flatten
model=Sequential()
model.add(Conv2D(filters=16, kernel_size=(5, 5), input_shape=(28, 28, 1), activation='relu'))
model.add(MaxPooling2D(pool_size=(2, 2)))
#model.add(Conv2D(filters=36, kernel_size=(5, 5), activation='relu'))
#model.add(MaxPooling2D(pool_size=(2, 2)))
model.add(Flatten())
model.add(Dense(10, activation='softmax'))
model.summary()
model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])
     
Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
=================================================================
 conv2d (Conv2D)             (None, 24, 24, 16)        416       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 12, 12, 16)       0         
 )                                                               
                                                                 
 flatten (Flatten)           (None, 2304)              0         
                                                                 
 dense (Dense)               (None, 10)                23050     
                                                                 
=================================================================
Total params: 23,466
Trainable params: 23,466
Non-trainable params: 0
_________________________________________________________________

mnist=keras.datasets.mnist
(train_data, train_label), (test_data, test_label) = mnist.load_data()
test_label_original=test_label #for crosstab
train_label = keras.utils.to_categorical(train_label)
test_label = keras.utils.to_categorical(test_label)
     

train_history=model.fit(train_data,train_label,epochs=10,batch_size=512)
     
Epoch 1/10
118/118 [==============================] - 5s 15ms/step - loss: 3.9564 - accuracy: 0.8308
Epoch 2/10
118/118 [==============================] - 2s 14ms/step - loss: 0.2603 - accuracy: 0.9589
Epoch 3/10
118/118 [==============================] - 2s 14ms/step - loss: 0.1189 - accuracy: 0.9740
Epoch 4/10
118/118 [==============================] - 2s 15ms/step - loss: 0.0754 - accuracy: 0.9816
Epoch 5/10
118/118 [==============================] - 2s 15ms/step - loss: 0.0511 - accuracy: 0.9863
Epoch 6/10
118/118 [==============================] - 2s 15ms/step - loss: 0.0386 - accuracy: 0.9893
Epoch 7/10
118/118 [==============================] - 2s 13ms/step - loss: 0.0295 - accuracy: 0.9914
Epoch 8/10
118/118 [==============================] - 2s 13ms/step - loss: 0.0237 - accuracy: 0.9928
Epoch 9/10
118/118 [==============================] - 1s 10ms/step - loss: 0.0199 - accuracy: 0.9940
Epoch 10/10
118/118 [==============================] - 1s 10ms/step - loss: 0.0171 - accuracy: 0.9948

score=model.evaluate(test_data, test_label)
print(score[1])
     
313/313 [==============================] - 1s 4ms/step - loss: 0.1323 - accuracy: 0.9774
0.977400004863739

#繪製正確性與損失立歷史圖
import matplotlib.pyplot as plt
import numpy as np

def show_train_history(train_history, train, validation):
    plt.plot(train_history.history[train])
    plt.plot(train_history.history[validation])
    plt.title('Train History')
    plt.ylabel('Train')
    plt.xlabel('Epoch')
    plt.legend(['train', 'validation'], loc='center right')
    plt.show()

show_train_history(train_history, 'accuracy', 'accuracy')
show_train_history(train_history, 'loss', 'loss')

# Confusion Matrix混淆矩陣
import pandas as pd
prediction = model.predict(test_data)
prediction_label=np.argmax(prediction,axis=1)
print(test_label.shape)
pd.crosstab(test_label_original, prediction_label, rownames=['label'], colnames=['predict'])
